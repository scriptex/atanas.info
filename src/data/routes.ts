export enum Routes {
	HOME = '/',
	ABOUT = '/about',
	TIMELINE = '/timeline',
	SKILLS = '/skills',
	PORTFOLIO = '/portfolio',
	STATS = '/stats',
	SLIDES = '/slides',
	VIDEOS = '/videos',
	ARTICLES = '/articles',
	CERTIFICATES = '/certificates',
	SOCIAL = '/social',
	BLOG = '/blog',
	MUSIC = '/music'
}
