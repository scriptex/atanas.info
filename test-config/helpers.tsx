import { act } from 'react-dom/test-utils';
import type { FC } from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '@scripts/shared';

type ComponentType = FC<Readonly<React.ComponentProps<any>>>;

export const test = async (Component: ComponentType): Promise<RenderResult> => {
	let result!: RenderResult;

	await act(async () => {
		result = await waitFor(() => render(<Component />));
	});

	return result;
};

export const snapshotTest = (Component: ComponentType, buttonSelector?: string): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			const { asFragment, container } = await test(Component);

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				await act(async () => {
					await container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			const { unmount, asFragment } = await test(Component);

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};

export const mockAudioContext = (): jest.Mock => {
	const mockConnect = jest.fn();

	const mockgetByteFrequencyData = jest.fn();

	const mockcreateMediaElementSource = jest.fn(() => ({
		connect: mockConnect
	}));

	const mockcreateAnalyser = jest.fn(() => ({
		connect: mockConnect,
		frequencyBinCount: [0, 1, 2],
		getByteFrequencyData: mockgetByteFrequencyData
	}));

	const mockcreateOscillator = jest.fn(() => ({
		channelCount: 2
	}));

	const mockChannelSplitterConnect = jest.fn(n => n);

	const mockcreateChannelSplitter = jest.fn(() => ({
		connect: mockChannelSplitterConnect
	}));

	const mockedAudioContext = jest.fn(() => ({
		createAnalyser: mockcreateAnalyser,
		createOscillator: mockcreateOscillator,
		createChannelSplitter: mockcreateChannelSplitter,
		createMediaElementSource: mockcreateMediaElementSource
	}));

	return mockedAudioContext;
};

export const mockFetch = (value: Record<string, string | boolean>): void => {
	global.fetch = jest.fn().mockImplementation(() =>
		Promise.resolve({
			json: () => Promise.resolve(value)
		})
	);
};
