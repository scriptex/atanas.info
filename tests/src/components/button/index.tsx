import { Button } from '@components';
import { Routes } from '@data/routes';
import { snapshotTest } from '@test-config/helpers';

snapshotTest(() => <Button<'link'> type="link" href={Routes.ABOUT} />);
snapshotTest(() => <Button<'button'> type="button"></Button>);
snapshotTest(() => (
	<Button<'anchor'> type="anchor" rel="noopener noreferrer" href="https://atanas.info" target="_blank" />
));
