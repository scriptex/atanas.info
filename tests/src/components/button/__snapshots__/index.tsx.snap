// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` component Should render the  component 1`] = `
<DocumentFragment>
  <a
    class="c-btn"
    href="/about"
  />
</DocumentFragment>
`;

exports[` component Should render the  component 2`] = `
<DocumentFragment>
  <button
    class="c-btn"
    type="button"
  />
</DocumentFragment>
`;

exports[` component Should render the  component 3`] = `
<DocumentFragment>
  <a
    class="c-btn"
    href="https://atanas.info"
    rel="noopener noreferrer"
    target="_blank"
  />
</DocumentFragment>
`;

exports[` component Should unmount the  component 1`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 2`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 3`] = `<DocumentFragment />`;
